lib/Monads/NonDetMonad_Sat.thy

```diff
@@ -137,4 +137,12 @@ lemma gets_exs_valid:
   "\<lbrace>(=) s\<rbrace> gets f \<exists>\<lbrace>\<lambda>r. (=) s\<rbrace>"
   by (rule exs_valid_gets)
 
+lemma exs_valid_assert_opt[wp]:
+  "\<lbrace>\<lambda>s. G \<noteq> None \<and> Q (the G) s\<rbrace> assert_opt G \<exists>\<lbrace>Q\<rbrace>"
```

Comment:
Should have realised that earlier in MM: the preferred form for option lemmas is to avoid `the`. It looks nicer now, because you can avoid the quantifier, but it can bite you later when the ~= None gets simplified separately from the rest and potentially dropped/cut off. So the canonical formulation for the precondition is `\Ex x. G = Some x /\ Q x` (I'm assuming the weaker forall/implies form is not true for exs_valid).

It'll lead to needing to use the lifting lemma for `Ex` at the application site, but we do have these.